### DISCUSSION TITLE
paymentHistory 테이블에 상품 이름 추가

### DISCUSSION THREAD

**Discussion by Kwonminwoo** [SCBJ-7/SCBJ-BE#186] [Announcements]
### 안건
- `paymentHistory` 테이블에 상품 이름(`product_name`)을 추가했습니다.

### 고민사항
-  카카오페이 결제 시 `product name`이 필요하여 상품 이름을 어떻게 할지 고민했습니다. 사용자 입장에서 호텔이름 + 객실 이름으로 정하는게 가장 이해가 쉬울 것 같았습니다. 따라서 상품명을 호텔 이름 + 객실 이름으로 정하게 되었습니다.
![image](https://github.com/SCBJ-7/SCBJ-BE/assets/34360434/0f7a840d-ac52-4db9-9051-b2e9975a6917)

- 문제의 발단입니다. 카카오페이 결제는 최초 결제 요청, 결제 승인 요청 두 가지로 분리되어 있습니다. 여기서 최종적으로 결제 승인 요청이 성공하면 판매자에게 판매 완료 알림을 보내기 위해 `product name `이 필요합니다. 
- 호텔 이름과 객실 이름을 조회하기 위해서는 join을 4번 해야 합니다.(`productId`만 아는 상황) 최초 요청에서는 해당 정보 이외의 정보가 필요하기에 어차피 `join`을 해야했지만, 결제 승인 요청에서는 해당 정보 외의 정보가 필요하지 않기에 `join` 4번이나 해서 가져오기엔 성능상 손해가 크다고 생각했습니다. 따라서 결제 시 정보 검증을 위해 레디스를 이용해 데이터를 저장하고 있던 곳에 함께 저장하고 승인 요청 시 조회하는 방법을 택했습니다.
![image](https://github.com/SCBJ-7/SCBJ-BE/assets/34360434/0c066a48-43d6-4d8a-9b83-e3969da0c94a)

- 하지만 결제 후 정산 시 또 한번 사용자에게 정산이 완료되었다는 알림을 보내줘야 했습니다. 마찬가지로 `product name`이 필요했습니다. 정산은 매일 자정에 일어나기에 해당 시간에 레디스에서 정보를 조회해서 사용할 수는 없었습니다. 따라서 호텔, 객실 정보를 가지고 오기 위해서 4번의 `join`을 하거나, `paymentHistory`에 상품 이름을 저장하는 방식 중 하나를 선택해야 했습니다. 
- 정산 시 하루동안 결제 된 상품 모두를 조회해야하는 부담이 큰 작업이라고 생각했습니다. 이 때 조회마다 `join`이 4번 일어난다면 성능이 크게 저하될 것 이라 판단하였습니다.
- 그에 비해 `paymentHistory`에 상품 이름을 넣는다면 조회의 부담을 줄일 수 있고 저장(`insert`)시 `build` 할 때 데이터를 하나 추가하는 작업 밖에 없기 때문에 성능 상 이점이 더 많다고 판단되었습니다. 따라서 `PaymentHistory` 테이블에 상품 이름을 저장하는 방식으로 결정하게 되었습니다.
### 팀원
@YurimYang

**Comment by wocjf0513** [comment]
명필이시네요!! 💯 
좋은 근거로 DB에 필드를 추가하는 것도 괜찮은 거 같아요!!
=> 정산은 매일 자정에 일어나기에 해당 시간에 레디스에서 정보를 조회해서 사용할 수는 없었습니다. 라는  말이 잘 이해가 안되서 추가 설명 가능하신가요

**Reply by YurimYang** [reply]
자정마다 레디스에서 조회를 하는 것이 아니라, 레디스는 캐시용 DB이기 때문에 데이터가 사라진다면 저희가 정산 알림을 해줄 때 PRODUCT NAME을 갖고 올 수 없게 되어서 DB로 넣는 방향으로 결정하게 됐습니다.

**Reply by wocjf0513** [reply]
레디스의 성질도 생각해서 데이터를 넣는 방향성 너무 좋네요!!

**Reply by Aleexender** [reply]
저는 오히려 반대로 Join을 아무리 많이 해도 쿼리 튜닝을 통해서 문제를 해결해내려고 생각했었어요!
